Add index relationship to case mock

proteusvacuum opened the pull request:
I think a `CaseRelationship` in this mock is actually known as a `CaseIndex` elsewhere. @czue, do you think this is worth changing, or is there a reason you called it a `CaseRelationship` that I'm not seeing?(https://github.com/dimagi/commcare-hq/pull/6147)

Either way, this extra line needs to get added for testing cases with different index relationships (i.e. extensions).

@sravfeyn

proteusvacuum committed:
Add index relationship to case mock

proteusvacuum committed:
Rename CaseRelationship -> CaseIndex

relationships -> indices

proteusvacuum committed:
Set default relationship to child

dimagimon commented on corehq/ex-submodules/casexml/apps/case/tests/test_factory.py:
```diff
@@ -24,12 +24,13 @@ class CaseStructureTest(SimpleTestCase):
         self.assertEqual(
-            {CaseRelationship.DEFAULT_RELATIONSHIP: (CaseRelationship.DEFAULT_RELATED_CASE_TYPE, parent_case_id)},
+            {CaseIndex.DEFAULT_RELATIONSHIP: (CaseIndex.DEFAULT_RELATED_CASE_TYPE, parent_case_id,
+                                                     CaseIndex.DEFAULT_RELATIONSHIP)},
```
E127 continuation line over-indented for visual indent

proteusvacuum committed:
lint [ci skip]

proteusvacuum committed:
allow cases to have many indices of the same type

`index` was returning a dict... which isn't strictly correct, since
in theory a case can have many indices of the same relationship.

proteusvacuum commented:
I ended up changing it... Mostly so that I could write a test where there was more than one index of the same type.

Wait for tests, if they ever start (!)...

czue commented:
i don't remember why i called it relationship. possibly so that it was clearer what type of object I was working with (sometimes searching for things like `index` or `indices` in the code gets hard)? 

happy with this change if you find it more clear. agree we should wait for tests.

proteusvacuum committed:
Allow use of identifiers

dimagimon commented on corehq/ex-submodules/casexml/apps/case/mock.py:
```diff
@@ -278,38 +280,42 @@ class CaseStructure(object):
     DEFAULT_RELATED_CASE_TYPE = 'default_related_case_type'
 
-    def __init__(self, related_structure=None, relationship=DEFAULT_RELATIONSHIP, related_type=None):
+    def __init__(self, related_structure=None, relationship=DEFAULT_RELATIONSHIP, related_type=None, identifier=None):
```
E501 line too long (118 > 115 characters)

czue commented:
:+1: to merge when tests pass - feel free to ping or self-merge then

proteusvacuum commented:
A better way of doing this, https://github.com/dimagi/commcare-hq/pull/8600
